### PR TITLE
using okHttp connector with cache and proxy for github

### PR DESCRIPTION
**Pull request by lanwen**
this can speed up fetching data form GH and reduce rate limits consuming

[Review on Reviewable](https://reviewable.io/reviews/jenkinsci/github-plugin/88)

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
How does cache impact the request results reliability? In the code there is no invalidation hooks for the cache, so I suspect the sensitive information (group membership, login status responses, etc.) may be cached and lead to mistakes in the case of the improper usage. Does it also allow to brute-force the oauth token w/o going to real GitHub checks?

**Comment by lanwen** [comment]
GH + okHttp do all work for results reliability (https://developer.github.com/v3/#conditional-requests). 

How can token be brute-forced? Think stealing token from credentials is much easer.

**Comment by KostyaSha** [comment]
@oleg-nenashev please learn github documentation ;)

File: src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+ * @author lanwen (Merkushev Kirill)
+ * @see GitHubServerConfig#loginToGithub()
+ */
+@Restricted(NoExternalUse.class)
+public class GitHubLoginFunction extends NullSafeFunction<GitHubServerConfig, GitHub> {
+
+    /**
+     * Capacity of cache for GitHub client in MB.
+     *
+     * Can be overridden as system property -Dorg.jenkinsci.plugins.github.GitHubPlugin.cache.size.mb=1 on jenkins jvm
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java]: No, don't use properties, they sucks. User can't find what and where configure.

**Comment by KostyaSha** [comment]
It also misses reconnection algorithm that i so tuned and tested :(

**Comment by oleg-nenashev** [comment]
@KostyaSha please study community behavior BKMs :P

File: src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java (hunk 1 above)
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java]: think user should not configure it until very big troubles

**Comment by KostyaSha** [comment]
@oleg-nenashev do you remember rtfm times? ;)

File: src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    /**
+     * Uses proxy if configured on pluginManager/advanced page
+     *
+     * @param apiUrl GitHub's url to build proxy to
+     *
+     * @return proxy to use it in connector
+     */
+    private Proxy getProxy(String apiUrl) {
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java]: `@Nonnull` i suppose

Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java]: think in private methods it can be ignored

Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java]: It for static analysing, nobody save you from npes in private methods :)

File: src/main/resources/org/jenkinsci/plugins/github/config/GitHubPluginConfig/config.groovy [hunk 3]
```diff
@@ -38,6 +38,10 @@ f.section(title: descriptor.displayName) {
             }
         }
 
+        f.entry(title: _("GitHub client cache size (Mb)"), field: "clientCacheSize") {
```
Comment (KostyaSha) [inline src/main/resources/org/jenkinsci/plugins/github/config/GitHubPluginConfig/config.groovy]: `MB`

File: src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java (hunk 2 above)
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/internal/GitHubLoginFunction.java]: It for extra checks for null when use something. If method private and used in one place - its overhead to annotate methods with nonnull

File: src/main/java/org/jenkinsci/plugins/github/config/GitHubPluginConfig.java [hunk 4]
```diff
@@ -55,6 +53,13 @@
     public static final String GITHUB_PLUGIN_CONFIGURATION_ID = "github-plugin-configuration";
 
     /**
+     * Default value in MB for client cache size
+     *
+     * @see #getClientCacheSize()
+     */
+    private static final int DEFAULT_CLIENT_CACHE_SIZE_MB = 20;
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/config/GitHubPluginConfig.java]: It should be in GitHubServerConfig

File: src/main/resources/org/jenkinsci/plugins/github/config/GitHubPluginConfig/help-clientCacheSize.html [hunk 5]
```diff
@@ -0,0 +1,7 @@
+<div>
+    Cache size in MB used by GitHub client. This can speed up fetching data form GH and reduce rate limits consuming.<br/>
+    GH + okHttp do all work for results reliability 
+    (<a href="https://developer.github.com/v3/#conditional-requests">Conditional-requests in GitHub documentation</a>)<br/>
+    
+    Set to <b>0</b> to disable this feature
```
Comment (KostyaSha) [inline src/main/resources/org/jenkinsci/plugins/github/config/GitHubPluginConfig/help-clientCacheSize.html]: doble `to`, maybe just `Set 0 to disable..`?

Comment (lanwen) [inline src/main/resources/org/jenkinsci/plugins/github/config/GitHubPluginConfig/help-clientCacheSize.html]: yep

**Pull request merged**